crates/typst-layout/src/inline/shaping.rs

```diff
@@ -795,6 +808,15 @@ fn shape_segment<'a>(
     let pos = buffer.glyph_positions();
     let ltr = ctx.dir.is_positive();
 
+    let char_in_coverage = |char_start| {
+        let char_end = text[char_start..]
+            .char_indices()
+            .nth(1)
+            .map(|(offset, _)| offset + char_start)
+            .unwrap_or(text.len());
+        coverage.map_or(true, |cov| cov.is_match(&text[char_start..char_end]))
+    };
```

Comment:
Perhaps it would be easier to use the existing `let c = text[cluster..].chars().next().unwrap();` and move it upwards and then check `coverage.map_or(true, |cov| cov.is_match(c.encode_utf8(&mut [0; 4])))`? I'm not 100% certain whether it's safe to move the let up, but in principle the clusters should all correspond to valid char indices...

Reply:
okay, I see now that you also use it below, so perhaps it's not better. not sure ...

Reply:
At first I tried the way you described but it make code below more complex, and it have performance issue. So I switched to the current `char_in_coverage` function.